Dominate & Lazarus Inject basic mobs. Lazarus Injected mobs don't fight each other.

Jacquerel opened the pull request:
## About The Pull Request

Fixes #72404 
The Lazarus Injector doesn't currently work on basic mobs, but should.

Problem:
The EMPed state of the Lazarus Injector is intended to make a revived mob hostile to everyone except you, including other mobs you have revived wtih an EMPed Lazarus Injector.
This is trivial to achieve for Simple Mobs which essentially all share the same AI, but I could not think of a single workable solution for Basic Mobs which don't, or at least any which didn't come with a tedious requirement to closely consider this niche item when programming any additional AI.

Solution:
Change the default behaviour of the Lazarus Injector so this is not a problem.
If all it does it make the mob loyal to you _and_ friendly to other mobs which are loyal to you, then it's pretty easy because we can just use the existing faction flags.

This is unambiguously a buff to using the item for nefarious purposes as now if you revive four ice drakes and fulton them onto the station they won't kill each other until only one is left, but is the only workable solution I could really think of.
A lot of the very dangerous mining fauna can't be dragged so transporting your army to the station still poses a question.

The alternate solution was just to replace the AI controller of any emp-revived basic mob with a "zombie" AI controller, but this has the problem that
A- It would now make things like cows and dogs into hostile creatures when they previously weren't.
B- It loses any interesting behaviour the mob previously had and for cases like Bileworms doesn't even make any sense (they'd try to walk and just get stuck in place).
This ultimately leads to needing to make bespoke versions for various mobs, which doesn't seem desirable from a maintainability standpoint.

As a side note it's still not a great idea to revive Bileworms _anyway_ as, their ability to move is tied to their ability to attack so once they don't have a target they will just kind of sit there and if they _do_ get a target their attempts to help you fight are difficult to distinguish from attempts to kill you... but at least being able to revive them makes it easier to make one sapient if you really want to trap a player's mind inside a body which is incapable of leaving lavaland.

Additional edit:
At Fikou's suggestion I've also added a sentience comparison proc to `mob/living` and removed some code duplication which dealt with this problem in the sentience/mind transfer potions, as well as added it to the Dominate spell.

## Why It's Good For The Game

This device is meant to revive mobs and it shouldn't be required for players to memorise an arbitrary list of which mobs it does and doesn't work on.
Especially as the goal is eventually that all simple mobs should be basic mobs.
This way of working is more intuitive, even if it is also stronger. I was surprised when I used EMPed injectors and my "new minions" just killed each other.

## Changelog

:cl:
fix: You can now revive 'basic mobs' with a Lazarus Injector, such as dogs, cows, axolotls, or carp.
fix: The same category of mobs can also now be effected by the Runic Golem Dominate spell.
fix: Basic Mobs will switch target if they can no longer attack their current target; meaning that if you become a Bileworm's friend it will stop attacking you.
balance: Mobs injected with the Lazarus Injector while it is EMPed will no longer attack other mobs revived by EMPed Lazarus Injectors. 
/:cl:

Jacquerel committed:
Maybe as close as I can get

Jacquerel committed:
Forgot to remove the old code

Jacquerel committed:
I was overwriting my own faction changes

Jacquerel committed:
Actually didn't need this line either

Jacquerel committed:
The order of these procs matters

Jacquerel committed:
Find Potential Targets now also removes an invalid current target.

Jacquerel committed:
Don't get fancy, setting the faction to just neutral makes more sense than copying the entire list.

Jacquerel committed:
Merge branch 'master' into back-for-more

Jacquerel committed:
Adds sentience type comparison helper, applies it to the three places sentience type is checked
Because compare_sentience_type helpfully returns false on non-basic-or-simple mobs this also can eliminate some other checks

Jacquerel committed:
Merge branch 'back-for-more' of https://github.com/Jacquerel/orbstation into back-for-more

Jacquerel committed:
Clarifies some checks for future developers

Jacquerel committed:
While I'm here let's desoul these into balloon alerts
Also fixes an oopsie with an inverted check

LemonInTheDark commented on code/datums/ai/basic_mobs/basic_ai_behaviors/targetting.dm:
```diff
@@ -7,14 +7,20 @@
 		CRASH("No target datum was supplied in the blackboard for [controller.pawn]")
 
-	potential_targets = hearers(vision_range, controller.pawn) - living_mob //Remove self, so we don't suicide
+	var/datum/weakref/weak_target = controller.blackboard[target_key]
+	var/atom/current_target = weak_target?.resolve()
+	if (targetting_datum.can_attack(living_mob, current_target))
+		finish_action(controller, succeeded = TRUE)
+		return
+
+	controller.blackboard[target_key] = null
```
what is this? block all target f inding if we have a target already?

LemonInTheDark replied:
is it just to cover for simple_find_target? does this effect anything else it shouldn't?

LemonInTheDark commented on code/modules/research/xenobiology/xenobiology.dm:
```diff
@@ -689,45 +689,38 @@
-	else
-		to_chat(user, span_notice("[dumb_mob] looks interested for a moment, but then looks back down. Maybe you should try again later."))
+	if(!LAZYLEN(candidates))
+		balloon_alert(user, "try again later!")
```
trading out to_chats?

LemonInTheDark commented on code/modules/research/xenobiology/xenobiology.dm:
```diff
@@ -752,31 +745,24 @@
 
 	if(job_banned)
-		to_chat(user, span_warning("Your mind goes blank as you attempt to use the potion."))
+		balloon_alert(user, "you're banned!")
```
sadge

Jacquerel commented:
@LemonInTheDark i’m replying on mobile and it’s being shit about letting me actually reply to specific comments so to summarise here:

I moved the check for “don’t bother finding a new target if you already have one” off the planning subtree and into the behaviour.  
I also added a clause for “if it is no longer valid for you to target this mob then lose your current target and replace it”, which is the main reason why I moved it. 
The AI standards doc said that if you’re changing blackboard keys (such as dropping a target you already had) you should usually be doing it in behaviours and controllers should only read values, this is a rule broken in a few places but i’m trying to minimise that in my own work. 
I think this should be safe because it’s behaviour we should want in all cases, if a mobs targeting behaviour changes or the properties of its target change then it should react accordingly.  

Also yeah I replaced some to_chats with balloons basically because none of that error reporting seems worth logging in the chat.  
The ban message certainly has less soul but the other one was completely useless as feedback, it literally did not tell you what the actual problem was. Might be less immersive now but it’s more useful.

Jacquerel commented:
Should say “if it’s no longer valid to target this target” rather than “target this mob” up there because it’s used for things which aren’t mobs but another thing i apparently can’t do on mobile is edit my posts.

Jacquerel committed:
Merge branch 'master' into back-for-more

Jacquerel commented:
I'm back now so I can address any further reviews made but I'm not sure there's anything outstanding?

Jacquerel committed:
Merge branch 'master' into back-for-more
